### PR TITLE
Feat #25150: Pose Graph MST initialization

### PR DESCRIPTION
**Pull request by 03kiko**
Implements an MST-based initialisation for pose graphs, as proposed in issue #25150. Both Prim’s and Kruskal’s algorithms were added to the 3D module. These receive a vector of node IDs and a vector of edges (each with source and target IDs and a weight), and return a vector with the resulting edges. These MST implementations treat edges as undirected internally, meaning users only need to provide one direction (A→B or B→A), and duplicates are handled automatically.

Additionally, a new pose graph initialisation method using MST (Prim) was implemented. It constructs the MST over the pose graph, then traverses it to reconstruct node poses. With this, users can call `poseGraph->initializePosesWithMST()` to create an initial solution for the pose graph problem.

A set of test cases validating the implementation was also included.

#### Notes
- The edge weight used in the MST for pose graphs is calculated as:
`weight = || translation || + λ * rotation_angle`,
where λ = 0.485 was determined empirically based on optimiser performance;
- Validated on [Sphere-a](https://lucacarlone.mit.edu/datasets/) pose graph, showing similar convergence behaviour with or without MST initialisation;
- Alternative weight formulas, such as the Mahalanobis distance formula, were also tested, but the current formula yielded better results.

#### Future Work
- Extend testing to more diverse pose graphs (currently limited to [Sphere-a](https://github.com/opencv/opencv_extra/blob/5.x/testdata/cv/rgbd/sphere_bignoise_vertex3.g2o) in opencv_extra)
- Explore adaptive tuning of the λ parameter for broader applicability.

Co-authored-by: @miguel1099 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by 03kiko** [comment]
Hi @asmorkalov,

We noticed that no reviewers have been assigned to this PR yet. Is there a specific reason for this?  Let us know if there's anything we can do to move this PR forward.

Thanks!

File: modules/3d/include/opencv2/3d/detail/mst.hpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+CV_EXPORTS std::vector<MSTEdge> buildMSTPrim(
+    const size_t numNodes,
+    const std::vector<MSTEdge>& edges,
+    size_t root = 0
+);
+
+// builds a MST using Kruskal's algorithm
+CV_EXPORTS std::vector<MSTEdge> buildMSTKruskal(
+    const size_t numNodes,
+    const std::vector<MSTEdge>& edges
```
Comment (asmorkalov) [inline modules/3d/include/opencv2/3d/detail/mst.hpp]: I propose to use single function as public API with algorithm type parameter like it's done in many places in OpenCV. Also we have internal MST implementation at least in multi-camera calibration. It makes sense to move it to main header, not details.

File: modules/3d/include/opencv2/3d/detail/mst.hpp [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace cv
+{
+namespace detail
+{
+
+// represents an edge in a graph for MST computation
+struct CV_EXPORTS MSTEdge
+{
+    size_t source, target;
```
Comment (asmorkalov) [inline modules/3d/include/opencv2/3d/detail/mst.hpp]: size_t is not binding friendly. I propose to use just int here. Also you can use CV_EXPORTS_W_SIMPLE for the structure.

**Comment by 03kiko** [comment]
Hi @asmorkalov, thanks for the feedback! Please let us know if everything looks good now or if there’s anything else you'd like us to adjust!

**Comment by 03kiko** [comment]
Hi @asmorkalov, could you trigger CI again? It seems to have failed to clone opencv_extra?

**Comment by 03kiko** [comment]
@asmorkalov Friendly reminder

File: modules/3d/include/opencv2/3d/mst.hpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+};
+
+enum MSTAlgorithm
+{
+    MST_PRIM = 0,
+    MST_KRUSKAL = 1
+};
+
+// builds a MST using the selected algorithm (Prim or Kruskal).
+CV_EXPORTS std::vector<MSTEdge> buildMST(
```
Comment (asmorkalov) [inline modules/3d/include/opencv2/3d/mst.hpp]: CV_EXPORTS_W

File: modules/3d/include/opencv2/3d/mst.hpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+#ifndef OPENCV_3D_DETAIL_MST_HPP
+#define OPENCV_3D_DETAIL_MST_HPP
```
Comment (asmorkalov) [inline modules/3d/include/opencv2/3d/mst.hpp]: remove _DETAIL

File: modules/3d/include/opencv2/3d/mst.hpp [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <vector>
+
+namespace cv
+{
+
+// represents an edge in a graph for MST computation
```
Comment (asmorkalov) [inline modules/3d/include/opencv2/3d/mst.hpp]: Please use doxygen notation for the documentation.

File: modules/3d/include/opencv2/3d/mst.hpp (hunk 3 above)
Comment (asmorkalov) [reply modules/3d/include/opencv2/3d/mst.hpp]: The graph may have 2 or more isolated pieces. I propose the following:
1. return status, e.g. bool
2. add `CV_OUT std::vector<MSTEdge>&` parameter for result.

Also the solution will be more friendly solution for Python and Java bindings.

File: modules/3d/src/mst.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+namespace cv
+{
+
+std::vector<cv::MSTEdge> buildMST(int numNodes,
+                                  const std::vector<cv::MSTEdge>& edges,
+                                  MSTAlgorithm algorithm,
+                                  int root)
+{
+    switch (algorithm)
```
Comment (opencv-alalek) [inline modules/3d/src/mst.cpp:148]: Please use `CV_TRACE_FUNCTION();` as the first statement of exported function.

**Comment by 03kiko** [comment]
Hi @asmorkalov, @opencv-alalek, we've just pushed a new version addressing your reviews. Please let me know if everything looks good or if there's anything you'd like us to adjust.

**Comment by 03kiko** [comment]
Hi @asmorkalov, could you please re-trigger the CI? The Docker pull failed. Thanks!

File: modules/3d/src/mst.cpp [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    {
+        case MST_PRIM:
+            result = buildMSTPrim(numNodes, inputEdges, resultingEdges, root);
+            break;
+        case MST_KRUSKAL:
+            result = buildMSTKruskal(numNodes, inputEdges, resultingEdges);
+            break;
+        default:
+            CV_LOG_INFO(NULL, "Invalid MST algorithm specified");
+            result = false;
```
Comment (asmorkalov) [inline modules/3d/src/mst.cpp]: It's definitely invalid user input. It should not be just ignored. We use CV_Assert / CV_Error and throw exception with it.

File: modules/3d/src/mst.cpp [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    return (result && resultingEdges.size() >= static_cast<size_t>(numNodes - 1));
```
Comment (asmorkalov) [inline modules/3d/src/mst.cpp]: It's the only place where false is returned, besides the obvious one. What happens if the MST cannot be built, e.g. the graph consists of 2 split sub-graphs? I would say, it should return false too.

**Comment by 03kiko** [comment]
Hi @asmorkalov, just pushed a new version addressing your comments! Let me know if this looks good to you.

File: modules/3d/src/rgbd/pose_graph.cpp [hunk 9]
```diff
@@ -472,6 +479,112 @@ bool PoseGraphImpl::isValid() const
...
+{
+    double translationNorm = cv::norm(e.pose.t);
+
+    cv::Matx33d R = e.pose.q.toRotMat3x3(cv::QUAT_ASSUME_UNIT);
+    cv::Vec3d rvec;
+    cv::Rodrigues(R, rvec);
+    double rotationAngle = cv::norm(rvec);
+
+    // empirically determined
+    double lambda = 0.485;
```
Comment (asmorkalov) [inline modules/3d/src/rgbd/pose_graph.cpp]: Looks like some data specific parameter. I propose to add lambda parameter with default value 0.485 to `initializePosesWithMST` to have opportunity to change it without OpenCV modification and rebuild. The value may be promoted to the weight calculator as parameter too.

File: modules/3d/test/test_mst.cpp [hunk 10]
```diff
@@ -0,0 +1,267 @@
...
+
+    std::vector<MSTEdge> primMST, kruskalMST;
+    bool resultPrim = buildMST(numNodes, edges, primMST, MST_PRIM, 0);
+    bool resultKruskal = buildMST(numNodes, edges, kruskalMST, MST_KRUSKAL, 0);
+
+    EXPECT_TRUE(resultPrim);
+    EXPECT_TRUE(resultKruskal);
+    EXPECT_EQ(primMST.size(), static_cast<size_t>(numNodes - 1));
+    EXPECT_EQ(kruskalMST.size(), static_cast<size_t>(numNodes - 1));
+}
```
Comment (asmorkalov) [inline modules/3d/test/test_mst.cpp:271]: Please add a couple of tests with negative scenarios:
- split graph of 2-3 sub-graphs
- one isolated node in the graph

File: modules/3d/test/test_pose_graph.cpp [hunk 11]
```diff
@@ -155,6 +155,109 @@ TEST(PoseGraph, sphereG2O)
...
+                of << "v " << d.x << " " << d.y << " " << d.z << std::endl;
+            }
+            size_t esz = pgWithMSTAndOptimizer->getNumEdges();
+            for (size_t i = 0; i < esz; i++)
+            {
+                size_t sid = pgWithMSTAndOptimizer->getEdgeStart(i), tid = pgWithMSTAndOptimizer->getEdgeEnd(i);
+                of << "l " << sid + 1 << " " << tid + 1 << std::endl;
+            }
+            of.close();
+        }
```
Comment (asmorkalov) [inline modules/3d/test/test_pose_graph.cpp]: We have saveMesh / savePointCloud for obj and ply i/o. Please use it: https://docs.opencv.org/5.x/da/d35/group____3d.html#gad565f2878b3d8c3f959ae7967a520bb2

File: modules/3d/test/test_pose_graph.cpp [hunk 12]
```diff
@@ -155,6 +155,136 @@ TEST(PoseGraph, sphereG2O)
...
+            Point3d d = init->getNodePose(id).translation();
+            of << "v " << d.x << " " << d.y << " " << d.z << std::endl;
+        }
+        size_t esz = init->getNumEdges();
+        for (size_t i = 0; i < esz; i++)
+        {
+            size_t sid = init->getEdgeStart(i), tid = init->getEdgeEnd(i);
+            of << "l " << sid + 1 << " " << tid + 1 << std::endl;
+        }
+        of.close();
```
Comment (asmorkalov) [inline modules/3d/test/test_pose_graph.cpp]: Looks like this code is not needed any more, as you added implementation with `saveMesh` below.

**Comment by 03kiko** [comment]
> We have saveMesh / savePointCloud for obj and ply i/o. Please use it: https://docs.opencv.org/5.x/da/d35/group____3d.html#gad565f2878b3d8c3f959ae7967a520bb2

> Looks like this code is not needed any more, as you added implementation with `saveMesh` below.

Hi @asmorkalov, you're right that part of the code is not needed anymore, and I just left it there for comparison (I will comment out the old code).

Also, please take a look at what I said in my last commit, which you may have missed, and it explains the whole situation better:
Hi, sorry for taking so long to respond.
I've replaced the custom .obj writing logic with a call to `cv::saveMesh`, as suggested. For reference, the previous version (which I've kept for comparison) explicitly wrote:
- `v x y z` — for each node in the pose graph (as 3D points);
- `l i j` — to represent edges as lines between node indices.
The functions you suggested don’t perfectly fit a pose graph use case:
- `cv::savePointCloud` only captures the nodes — it doesn’t include any edge/connection structure;
- `cv::saveMesh` (used in the current commit) seems to be meant for triangular or polygonal meshes, and thus writes `f i j k` (faces) instead of `l i j` (lines) (in the commit we are only passing it 2 vertices, making it write f i j only). As a workaround, we could pass a fake third index by duplicating one vertex (e.g., f i j i to mimic a line), but most 3D viewers recognise these as degenerate faces and don’t render them correctly.

So I’d like to ask for your opinion on the best approach here:
- Keep the original version with explicit .obj line writing (v, l);
- Continue using cv::saveMesh, feeding 2 or 3 vertex indices (f i j or f i j i), although the pose graphs won't be visible in most 3d viewers;
- Use the Mesh struct already implemented in test_pose_graph.cpp (around line ~290);
- ... or any other direction you’d recommend.

Let me know what you think!

Note: We are using this [3d viewer](https://3dviewer.net/) to open the .obj files and see the pose graph(s).

**Comment by asmorkalov** [comment]
Let's presume the original "hand made" version, but please add a comment why it differs from saveMesh. It's not obvious without detailed code analysis.

**Review by asmorkalov — APPROVED** [review]

**Comment by 03kiko** [comment]
Hi @asmorkalov,

Thank you for approving our PR. Could you please merge it when you have a moment? Let us know if there's anything else you need from us.
